File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        if not root:
            return
        L = []
        heapq.heapify(L)

        def DFS(root):
            if not root:
                return
            DFS(root.left)
            heapq.heappush(L, root.val)
            DFS(root.right)
        DFS(root)
        return heapq.nsmallest(k, L).pop()

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void DFS(TreeNode* root, priority_queue<int, vector<int>, greater<int>>& q) {
        if (!root) {
            return;
        }
        DFS(root->left, q);
        q.push(root->val);
        DFS(root->right, q);
    }

    int kthSmallest(TreeNode* root, int k) {
        priority_queue<int, vector<int>, greater<int>> lQ;

        DFS(root, lQ);
        for (int i = 0; i < k - 1; i++) {
            lQ.pop();
        }

        return lQ.top();
    }
};